hack/generator/pkg/astmodel/type_test.go

```diff
@@ -0,0 +1,44 @@
+/*
+ * Copyright (c) Microsoft Corporation.
+ * Licensed under the MIT license.
+ */
+
+package astmodel
+
+import (
+	. "github.com/onsi/gomega"
+	"testing"
+)
+
+/*
+ * TypeAsObjectType() tests
+ */
+
+func TestTypeAsObjectType_GivenWrappedType_UnwrapsExpectedObject(t *testing.T) {
+
+	objectType := NewObjectType()
+	flagged := ArmFlag.Wrap(objectType)
+	optional := NewOptionalType(objectType)
+
+	cases := []struct {
+		name    string
+		subject Type
+		object  *ObjectType
+	}{
+		{"Object is", objectType, objectType},
+		{"Flagged type unwraps", flagged, objectType},
+		{"Optional type unwraps", optional, objectType},
+	}
+
+	for _, c := range cases {
+		c := c
+		t.Run(c.name, func(t *testing.T) {
+			t.Parallel()
+			g := NewGomegaWithT(t)
+			ot, err := TypeAsObjectType(c.subject)
```

Comment:
I'm not sure that this is what we want actually.

Right now practically speaking the `TypeAsObjectType` function only collapses ARM types. I know there's code in there to collapse `Optional` types too but even that I think is actually not correct (in the context of its use). I don't think the `Optional` path is taken right now because this is always called on a `definition` and `definitions` aren't ever to optional types (only properties are).

The problem I see here is that collapsing _all_ flags when this is called means that no flag can persist past the ARM type conversion phase (or any phase that makes use of this function).

I _think_ we want to allow preservation of flags all the way until the time we're writing code to disk (maybe even past that if we have post-writing phases?)...

It's late on a Friday and I'm maybe a bit crazy but is the problem that we have this desire to do stuff to `Object` (examine it, modify it, etc) and it's difficult to do that when it's wrapped? Maybe we need to rewrite the ARM to use a visitor so that it can preserve the wrapping types? (I think I wrote it originally before Visitor was added and then thought a bit about refactoring it but it melted my brain at the time). 

Or maybe we need to write some psudeo-monadic functions on FlagType that can do "stuff" with the inner type but all without leaving the context of the wrapper? (@Porges)

It's possible you fix this in a future PR and if so feel free to ignore this

Reply:
I _think_ `TypeAsObjectType()` is correct for our current usage, but I also see where you're coming from. I'll skip doing this here, but will lodge an issue so we don't forget to circle back and eliminate it.

Reply:
This seems an ok course of action.  Will leave this active though as I am curious about @porges and @babbageclunk's thoughts on a pattern that can achieve what this function wants to achieve but in a way that is more amenable to "other types"